src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/TagList.cs

```diff
@@ -0,0 +1,553 @@
+        public TagList(ReadOnlySpan<KeyValuePair<string, object?>> tagList) : this()
+        {
+            _tagsCount = tagList.Length;
+            switch (_tagsCount)
+            {
+                case 8:
+                    Tag8 = tagList[7];
+                    goto case 7;
+
+                case 7:
+                    Tag7 = tagList[6];
+                    goto case 6;
+
+                case 6:
+                    Tag6 = tagList[5];
+                    goto case 5;
+
+                case 5:
+                    Tag5 = tagList[4];
+                    goto case 4;
+
+                case 4:
+                    Tag4 = tagList[3];
+                    goto case 3;
+
+                case 3:
+                    Tag3 = tagList[2];
+                    goto case 2;
+
+                case 2:
+                    Tag2 = tagList[1];
+                    goto case 1;
+
+                case 1:
+                    Tag1 = tagList[0];
+                    break;
+
+                case 0: return;
+
+                default:
+                    Debug.Assert(_tagsCount > 8);
+                    _overflowTags = new KeyValuePair<string, object?>[_tagsCount + 8]; // Add extra 8 slots for more tags to add if needed
+                    tagList.CopyTo(_overflowTags);
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Gets the number of tags contained in the <see cref="T:System.Diagnostics.TagList" />.
+        /// </summary>
+        public readonly int Count => _tagsCount;
+
+        /// <summary>
+        /// Gets a value indicating whether the <see cref="T:System.Diagnostics.TagList" /> is read-only. This property will always return <see langword="false" />.
+        /// </summary>
+        public readonly bool IsReadOnly => false;
+
+        /// <summary>
+        /// Gets or sets the tags at the specified index.
+        /// </summary>
+        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="index" /> is not a valid index in the <see cref="T:System.Diagnostics.TagList" />.</exception>
+        public KeyValuePair<string, object?> this[int index]
+        {
+            readonly get
+            {
+                if ((uint)index >= (uint)_tagsCount)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(index));
+                }
+
+                if (_overflowTags is not null)
+                {
+                    Debug.Assert(index < _overflowTags.Length);
+                    return _overflowTags[index];
+                }
+
+                return index switch
+                {
+                    0 => Tag1,
+                    1 => Tag2,
+                    2 => Tag3,
+                    3 => Tag4,
+                    4 => Tag5,
+                    5 => Tag6,
+                    6 => Tag7,
+                    7 => Tag8,
+                    _ => default, // we shouldn't come here anyway.
```

Comment:
I hesitate to throw exception here because it wouldn't be right callers get exception for something not their fault. I'll add assert here instead. Thanks!

Reply:
Yeah, it's not their fault but it doesn't leave them in a good state. I'm not strongly advocating for, but debug.assert would only happen on our tests.

Reply:
Our tests covering such cases and we run debug builds in the CI. There where we catch such issues.